Enhance `zend_dump_op_array` to Properly Represent Non-Printable Characters (GH-15680)

WangYihang opened the pull request:
This pull request addresses issue GH-15680 which proposed an enhancement for `zend_dump_op_array` to properly represent non-printable characters in strings. 

This is useful for debugging purposes, as it allows developers to see the actual content of strings that contain non-printable characters.

I have also added testcases for this change, and all tests related to this pull request were **passed**.

For example, for the following php code:

```php
<?php
fwrite($buffer, "\x00\x01\xff--\n");
```

Current Output:

```c
0040 INIT_FCALL 2 112 string("fwrite")
0041 SEND_VAR CV6($buffer) 1
0042 SEND_VAL string("--
") 2                                     <------ new line (\n) is directly printed
0043 DO_ICALL
Expected Output:
```

With this change, the output might looks like:

```c
0040 INIT_FCALL 2 112 string("fwrite")
0041 SEND_VAR CV6($buffer) 1
0042 SEND_VAL string("\x00\x01\xff--\n") 2           <------ properly escaped with readable representations
0043 DO_ICALL
```

Related Issues/Pull requests:

+ https://github.com/php/php-src/issues/15680
+ https://github.com/php/php-src/pull/11337
+ https://github.com/php/php-src/issues/10576

dstogov commented on ext/standard/string.c:
```diff
@@ -3863,6 +3878,55 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
 /* }}} */
 
+static const char *get_char_repr(char c) {
+    for (size_t i = 0; i < sizeof(char_reprs) / sizeof(char_reprs[0]); i++) {
+        if (char_reprs[i].c == c) {
+            return char_reprs[i].repr;
+        }
+    }
```
This loop is executed for every character of the input string. It makes sense to collect the necessary symbols (similar to ``php_charmask()`` or use constant bitmask like https://github.com/php/php-src/blob/master/ext/json/json_encoder.c#L390) to make the following tests cheaper.

WangYihang replied:
Thanks for your valuable feedback! 

Can we just hardcode the char mapping in the code since it is a static mapping relation which might not need to be recalculated everytime user calls `php_repr_str`?

The hardcoded approach might bloat the code by adding 256 entries in this file.

WangYihang replied:
Solved by adding a static mapping array. 

I also conducted a simple performance (just count the CPU ticks) comparison. The following result shows the new version have **saved 32.56% CPU time**.

```
performance comparison  before        after
number of calls:        1000000       1000000
round of tests:         10            10
cpu ticks avg:          358170        241565 (-32.56%)
cpu ticks median:       356033        241496 (-32.17%)
cpu ticks stdev:        6564          372    (-94.33%)
```

Thank you so much for pointing out this performance considerations. I have learned a lot. Thanks! By the way, is there any performance benchmarking tool / framework for the Zend engine? I didn't find how to benchmark the performance so I wrote a simple c program to collect and compare the results. It would be better to have an automated benchmarking framework.

```c
zend_string *str = zend_string_init("Hello World", strlen("Hello World"), 0);
long usedTime = 0;
long numTests = 10;
int rounds = 1000000;
long *timeArray = malloc(numTests * sizeof(long));

for (int i = 0; i < numTests; i++) {
    usedTime = 0;
    for (int j = 0; j < rounds; j++) {
        long startTime = clock();
        zend_string *escaped_string = php_repr_str((str)->val, (str)->len);
        long endTime = clock();
        usedTime += endTime - startTime;
        zend_string_release(escaped_string);
    }
    timeArray[i] = usedTime;
}
```

WangYihang replied:
@dstogov Please check it out if you have time, thanks a lot!

dstogov reviewed (commented):
See the optimisation suggestion.
The rest is fine.

nielsdos commented on ext/standard/string.c:
```diff
@@ -3863,6 +3878,55 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
+        }
+    }
+    
+    repr_str = emalloc(new_len + 1);
```
Minor suggestion:
Now you do emalloc+zend_string_init. You can already allocate a zend_string here by using zend_string_alloc. This will give you a `zend_string *` and you can then use `ZSTR_VAL` to write the characters directly into here. It avoids copying the string.

WangYihang replied:
Cool! That's a great idea, I would like to allocate the maximum required size and then shrink it after the encoding process just like what `php_addcslashes` does.

WangYihang replied:
Done, thank you for your feedback! Just like this reply (https://github.com/php/php-src/pull/15730#discussion_r1742659604). I have done some performance benchmarking, it turns out that the new version outperforms the old version by ~32% in CPU time.

WangYihang replied:
https://github.com/php/php-src/pull/15730/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R3869-R3897

dstogov commented on ext/standard/charrepr.h:
```diff
@@ -0,0 +1,284 @@
+    const unsigned char len;
+} char_repr_t;
+
+static const char_repr_t char_reprs[256] = {
```
What is the reason to separate this array into a header?
It's going to be duplicated every time the header is included.

WangYihang replied:
The reason I separated this array into a header file was primarily for **modularity** and **reusability**. 

The current file (`string.c`) is already quite long, and combining these contents with the C file might make it harder to read and maintain. 

However, if you believe this separation is unnecessary, I can certainly move this part back into `string.c`.

iluuu1994 replied:
Was this code copied from somewhere? There's no header/license that would indicate so.

dstogov reviewed (commented):
I don't see big problems, but I also don't see a real need for this.
@iluuu1994 what do you think?

WangYihang commented on ext/standard/string.c:
```diff
@@ -3863,6 +3866,38 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
+		target += repr_len;
+    }
+	// add the trailing quote
+	*target++ = '"';
```
Fixed in https://github.com/php/php-src/pull/15730/commits/90fe73f94929f9f96ed738fddf445f1febb54cf8.

WangYihang commented:
> I don't see big problems, but I also don't see a real need for this. @iluuu1994 what do you think?

Thank you for your feedback.

While I understand that this enhancement might not seem immediately necessary, I believe it could bring several significant benefits:

+ Improved Developer Experience: Introducing a dedicated function or enhancing the existing one would provide developers with a more reliable way to handle non-printable characters, reducing the risk of errors in edge cases and improving code readability and maintainability. This need has been raised by several developers concerning string representation in PHP. Similar work has already been merged ([1] [2]), which focused on quoting a few specific characters like newlines. My pull request builds on that foundation by extending the functionality to cover all non-printable characters, making it a logical next step that should be considered for inclusion.

+ Consistency Across Languages: Many modern programming languages, such as Python (`repr`), Ruby (`inspect`), and JavaScript (`JSON.stringify`), offer similar functionality. Having a comparable feature in PHP would align it more closely with these languages, providing a familiar toolset for developers who work across multiple environments.

+ Prevention of Display Issues: During debugging, dumping non-printable control characters can disrupt the terminal's display, making the output difficult to interpret. A function that safely represents these characters would prevent such issues, improving debugging efficiency.

+ Easier Debugging with `var_dump`: Currently, PHP's `var_dump` function does not quote strings, which can make it challenging for developers to distinguish between visible and non-visible characters during debugging. This can lead to confusion and increased difficulty in identifying issues within the code. A standardized way to represent strings, including all non-printable characters, would alleviate this problem. Which could be left for the future work (If you guys agreed that it worthwhile, I can help to implement that).

+ Future-Proofing: Even if the need isn’t pressing at this moment, implementing this feature could support future debugging and logging functionalities, helping to keep PHP robust and user-friendly as it evolves.

I believe this change, while relatively minor, could add substantial value by addressing these points. I’m open to further refining the implementation based on your feedback.

What are your thoughts on this? @iluuu1994 

+ https://github.com/php/php-src/pull/11337
+ https://github.com/php/php-src/issues/10576

dstogov commented:
@iluuu1994 @nielsdos @arnaud-lb please make a decision.

iluuu1994 commented on ext/standard/charrepr.h:
```diff
@@ -0,0 +1,284 @@
+	{"\\\"", 2},
+	{"#", 1},
+	{"$", 1},
+	{"%%", 1},
```
This looks fishy. `%%` only needs to be escaped in `printf`. Given the impl uses `memcpy` rather than `strcpy`, it shouldn't lead to a bug though.

WangYihang replied:
> Was this code copied from somewhere? There's no header/license that would indicate so.

Actually, I just generated this header file by a few lines of ad-hoc Python code.

WangYihang replied:
> This looks fishy. %% only needs to be escaped in printf. Given the impl uses memcpy rather than strcpy, it shouldn't lead to a bug though.

Ahhhh, yes, I will fix it soon.

WangYihang replied:
@iluuu1994 Fixed in 32340baf165f7ec8d0391fa6bdcb37eca3791403.

https://github.com/php/php-src/pull/15730/commits/32340baf165f7ec8d0391fa6bdcb37eca3791403#diff-9dc13ac25ae79e9c7bd257341e953577aae1b2cb7562fb9acd64ce06c1baa71eL63-R63

iluuu1994 reviewed (approved):
I don't object to changing this. LGTM otherwise.

iluuu1994 commented on ext/standard/string.c:
```diff
@@ -3863,6 +3866,38 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
 /* }}} */
 
+/* {{{ php_repr_str */
+PHPAPI zend_string *php_repr_str(const char *str, size_t len) {
```
Ah, and I would also improve the naming. I don't find this particularly descriptive.

WangYihang replied:
Any suggestions about the naming? basically this function take a string as input, returns a zend_string which converts all non-printable chars to a human readable format (e.g. `chr(256)` will be convert to `"\xff"`).

The `repr` part in the function name was inspired by the python [`repr` function](https://docs.python.org/3/library/functions.html#repr).

> `repr` returns a string containing a printable representation of an object in Python.

arnaud-lb commented on ext/standard/string.c:
```diff
@@ -3863,6 +3866,38 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
 /* }}} */
 
+/* {{{ php_repr_str */
+PHPAPI zend_string *php_repr_str(const char *str, size_t len) {
```
I think this function has the same output as `smart_str_append_escaped()` (minus the enclosing double quotes). Is there anything blocking us from using it?

arnaud-lb commented on Zend/Optimizer/zend_dump.c:
```diff
@@ -67,10 +67,8 @@ void zend_dump_const(const zval *zv)
-
-			fprintf(stderr, " string(\"%s\")", ZSTR_VAL(escaped_string));
-
+			zend_string *escaped_string = php_repr_str(Z_STR_P(zv)->val, Z_STR_P(zv)->len);
+			fprintf(stderr, " string(%s)", ZSTR_VAL(escaped_string));
```
We can achieve a similar result with `php_addcslashes(Z_STR_P(zv), "\x00..\x1f\\\\\x7f..\xff", 2)` (with the different that it uses an octal representation), or the same result with `smart_str_append_escaped()`.

arnaud-lb reviewed (commented):
I like this change. This looks good to me, but I think we could reuse some existing code.
